sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1359 @@
+    transformation_ : array, shape (n_features_out, n_features)
+        The linear transformation learned during fitting.
+
+    n_neighbors_ : int
+        The provided ``n_neighbors`` is decreased if it is greater than or
+        equal to  min(number of elements in each class).
+
+    n_iter_ : int
+        Counts the number of iterations performed by the optimizer.
+
+    opt_result_ : scipy.optimize.OptimizeResult (optional)
+        A dictionary of information representing the optimization result.
+        This is stored only if ``store_opt_result`` is True.
+
+    Examples
+    --------
+    >>> from sklearn.neighbors import LargeMarginNearestNeighbor
+    >>> from sklearn.neighbors import KNeighborsClassifier
+    >>> from sklearn.datasets import load_iris
+    >>> from sklearn.model_selection import train_test_split
+    >>> X, y = load_iris(return_X_y=True)
+    >>> X_train, X_test, y_train, y_test = train_test_split(X, y,
+    ... stratify=y, test_size=0.7, random_state=42)
+    >>> lmnn = LargeMarginNearestNeighbor(n_neighbors=3, random_state=42)
+    >>> lmnn.fit(X_train, y_train) # doctest: +ELLIPSIS
+    LargeMarginNearestNeighbor(...)
+    >>> # Fit and evaluate a simple nearest neighbor classifier for comparison
+    >>> knn = KNeighborsClassifier(n_neighbors=3)
+    >>> knn.fit(X_train, y_train) # doctest: +ELLIPSIS
+    KNeighborsClassifier(...)
+    >>> print(knn.score(X_test, y_test))
+    0.933333333333
+    >>> # Now fit on the data transformed by the learned transformation
+    >>> knn.fit(lmnn.transform(X_train), y_train) # doctest: +ELLIPSIS
+    KNeighborsClassifier(...)
+    >>> print(knn.score(lmnn.transform(X_test), y_test))
+    0.971428571429
+
+    Notes
+    -----
+    Large margin nearest neighbor (LMNN) is a machine learning algorithm for
+    metric learning. It learns a (pseudo-)metric in a supervised fashion to
+    improve the classification accuracy of the k-nearest neighbor rule.
+    The main intuition behind LMNN is to learn a pseudometric under which all
+    data instances in the training set are surrounded by at least k instances
+    that share the same class label. If this is achieved, the leave-one-out
+    error is minimized.
+    This implementation follows closely Kilian Weinberger's MATLAB code found
+    at <https://bitbucket.org/mlcircus/lmnn> which solves the unconstrained
+    problem, finding a linear transformation with L-BFGS instead of solving the
+    constrained problem that finds the globally optimal metric. Different from
+    the paper, the problem solved by this implementation is with the squared
+    hinge loss (to make the problem differentiable).
+
+    .. warning::
+
+        Exact floating-point reproducibility is generally not guaranteed (
+        unless special care is taken with library and compiler options). As
+        a consequence, the transformations computed in 2 identical runs of
+        LargeMarginNearestNeighbor can differ from each other. This can
```

Comment:
But surely it's still deterministic under a fixed random state on a fixed architecture?

Reply:
Not even that is for sure, if one uses PCA for initialization. This goes back to `scipy.linalg.svd` and the warning is based on the [answer I got from scipy devs](https://github.com/scipy/scipy/issues/7878) . The transformation could not always be reproduced in 32bit systems, but it could theoretically also fail on 64bit, although it's very unlikely.